merkle/src/tree_impl.rs

```diff
@@ -0,0 +1,182 @@
+            } else {
+                let left = index * 2 + 1;
+                let right = index * 2 + 2;
+                let left_hash = self.calculate_hash_if_needed(left);
+                let right_hash = self.calculate_hash_if_needed(right);
+                let hash = TMerger::merge(&left_hash, &right_hash);
+                self.nodes[index] = hash.clone();
+                hash.map(|hash| (hash, MerkleTreeNodeMetadata::new(index)))
+            }
+        } else {
+            let leaf_index = index - self.nodes.len();
+            if leaf_index < self.leafs.len() {
+                let (hash, _) = &self.leafs[leaf_index];
+                Some((hash.clone(), MerkleTreeNodeMetadata::new(index)))
+            } else {
+                None
+            }
+        }
+    }
+}
+
+impl<TItem, THash, THasher, TMerger> MerkleTree for MinaMerkleTree<TItem, THash, THasher, TMerger>
+where
+    THasher: MerkleHasher<Item = TItem, Hash = THash>,
+    TMerger: MerkleMerger<Hash = THash>,
+    THash: Clone,
+{
+    type Item = TItem;
+    type Hash = THash;
+
+    fn depth(&self) -> u32 {
+        self.depth
+    }
+
+    fn count(&self) -> usize {
+        self.leafs.len()
+    }
+
+    fn root(&mut self) -> Option<Self::Hash> {
+        self.calculate_hash_if_needed(0).map(|(hash, _)| hash)
+    }
+
+    fn add_batch(&mut self, items: Vec<Self::Item>) {
+        let new_leaf_count = self.leafs.len() + items.len();
+        let new_depth = calculate_depth(new_leaf_count);
+        if new_depth != self.depth {
+            let new_node_count = calculate_node_count(new_depth);
+            self.depth = new_depth;
+            self.nodes = vec![None; new_node_count];
+        } else {
+            let start = self.nodes.len() + self.leafs.len();
+            for i in start..(start + items.len()) {
+                self.clear_dirty_hashes(i);
+            }
+        }
+        for leaf in items
+            .into_iter()
+            .enumerate()
+            .map(|(i, item)| (THasher::hash(&item, MerkleTreeNodeMetadata::new(i)), item))
+        {
+            self.leafs.push(leaf);
+        }
+    }
+
+    fn add(&mut self, item: Self::Item) {
+        self.add_batch(vec![item])
+    }
+}
+
+impl<TItem, THash, THasher, TMerger> Default for MinaMerkleTree<TItem, THash, THasher, TMerger>
+where
+    THasher: MerkleHasher<Item = TItem, Hash = THash>,
+    TMerger: MerkleMerger<Hash = THash>,
+    THash: Clone,
+{
+    fn default() -> Self {
+        Self {
+            depth: 0,
+            leafs: Vec::new(),
+            nodes: Vec::new(),
+            _pd_hasher: Default::default(),
+            _pd_merger: Default::default(),
+        }
+    }
+}
+
+fn calculate_depth(size: usize) -> u32 {
+    if size < 2 {
+        0
+    } else {
+        (size as f64).log2().ceil() as u32
+    }
+}
+
+fn calculate_node_count(depth: u32) -> usize {
+    2_usize.pow(depth) - 1
+}
+
+fn calculate_parent_index(index: usize) -> usize {
+    (index - 1) / 2
```

Comment:
`index.saturating_sub(1)` as `index-1` will panic, if `index` is `0`

Reply:
This is a private function and the index is not supposed to be 0. If that ever happens I guess it's better to panic than returning unexpected values.

Reply:
Let me add 'debug_assert!(index > 0);' here